drivers/adc/adc_lmp90xxx.c

```diff
@@ -0,0 +1,1136 @@
+		i++;
+	}
+
+	/* Instruction Byte 2 */
+	inst2 |= lmp90xxx_inst2_sz(len);
+	tx_buf[i].buf = &inst2;
+	tx_buf[i].len = sizeof(inst2);
+	dummy += sizeof(inst2);
+	i++;
+
+	/* Dummy RX Bytes */
+	rx_buf[0].buf = NULL;
+	rx_buf[0].len = dummy;
+
+	/* Data Byte(s) */
+	rx_buf[1].buf = dptr;
+	rx_buf[1].len = len;
+
+	tx.buffers = tx_buf;
+	tx.count = i;
+	rx.buffers = rx_buf;
+	rx.count = 2;
+
+	err = spi_transceive(data->spi_dev, &data->spi_cfg, &tx, &rx);
+	if (!err) {
+		data->ura = ura;
+	} else {
+		/* Force INST1 + UAB on next access */
+		data->ura = LMP90XXX_INVALID_URA;
+	}
+
+	k_mutex_unlock(&data->ura_lock);
+
+	return err;
+}
+
+static int lmp90xxx_read_reg8(struct device *dev, u8_t addr, u8_t *val)
+{
+	return lmp90xxx_read_reg(dev, addr, val, sizeof(val));
+}
+
+static int lmp90xxx_write_reg(struct device *dev, u8_t addr, u8_t *dptr,
+			      size_t len)
+{
+	struct lmp90xxx_data *data = dev->driver_data;
+	u8_t ura = LMP90XXX_URA(addr);
+	u8_t inst1_uab[2] = { LMP90XXX_INST1_WAB, ura };
+	u8_t inst2 = LMP90XXX_INST2_WB | LMP90XXX_LRA(addr);
+	struct spi_buf tx_buf[3];
+	struct spi_buf_set tx;
+	int i = 0;
+	int err;
+
+	if (len == 0) {
+		LOG_ERR("attempt write 0 bytes to register 0x%02x", addr);
+		return -EINVAL;
+	}
+
+	k_mutex_lock(&data->ura_lock, K_FOREVER);
+
+	if (ura != data->ura) {
+		/* Instruction Byte 1 + Upper Address Byte */
+		tx_buf[i].buf = inst1_uab;
+		tx_buf[i].len = sizeof(inst1_uab);
+		i++;
+	}
+
+	/* Instruction Byte 2 */
+	inst2 |= lmp90xxx_inst2_sz(len);
+	tx_buf[i].buf = &inst2;
+	tx_buf[i].len = sizeof(inst2);
+	i++;
+
+	/* Data Byte(s) */
+	tx_buf[i].buf = dptr;
+	tx_buf[i].len = len;
+	i++;
+
+	tx.buffers = tx_buf;
+	tx.count = i;
+
+	err = spi_write(data->spi_dev, &data->spi_cfg, &tx);
+	if (!err) {
+		data->ura = ura;
+	} else {
+		/* Force INST1 + UAB on next access */
+		data->ura = LMP90XXX_INVALID_URA;
+	}
+
+	k_mutex_unlock(&data->ura_lock);
+
+	return err;
+}
+
+static int lmp90xxx_write_reg8(struct device *dev, u8_t addr, u8_t val)
+{
+	return lmp90xxx_write_reg(dev, addr, &val, sizeof(val));
+}
+
+static int lmp90xxx_soft_reset(struct device *dev)
+{
+	int err;
+
+	err = lmp90xxx_write_reg8(dev, LMP90XXX_REG_RESETCN,
+				  LMP90XXX_REG_AND_CNV_RST);
+	if (err) {
+		return err;
+	}
+
+	/* Write to RESETCN twice in order to reset mode as well as registers */
+	return lmp90xxx_write_reg8(dev, LMP90XXX_REG_RESETCN,
+				   LMP90XXX_REG_AND_CNV_RST);
+}
+
+static inline bool lmp90xxx_has_channel(struct device *dev, u8_t channel)
+{
+	const struct lmp90xxx_config *config = dev->config->config_info;
+
+	if (channel >= config->channels) {
+		return false;
+	} else {
+		return true;
+	}
+}
+
+static inline bool lmp90xxx_has_input(struct device *dev, u8_t input)
+{
+	const struct lmp90xxx_config *config = dev->config->config_info;
+
+	if (input >= LMP90XXX_MAX_INPUTS) {
+		return false;
+	} else if (config->channels < LMP90XXX_MAX_CHANNELS &&
+		   (input >= 3 && input <= 5)) {
```

Comment:
All AFEs with the maximum number of channels have inputs 0-7, and the other ones only 0-2 and 6-7, right? Maybe some comment could help make this rule more visible. Or is it something obvious?

Reply:
Correct. It should be obvious from the data sheet and especially from the pinout of the given device.

Reply:
I have added a comment to the code which should hopefully make it more clear.